File: app/src/test/java/com/lucasurbas/counter/service/RunningCounterPresenterTest.java
package com.lucasurbas.counter.service;

import com.lucasurbas.counter.data.model.Counter;
import com.lucasurbas.counter.service.usecase.GetCounterInteractor;
import com.lucasurbas.counter.service.usecase.GetRunningCountersUpdatesInteractor;
import com.lucasurbas.counter.service.usecase.UpdateCounterInteractor;
import com.lucasurbas.counter.ui.explore.mapper.UiCounterItemMapper;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.subjects.PublishSubject;

import static junit.framework.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class RunningCounterPresenterTest {

    private static final int ID_1 = 1;
    private static final Counter COUNTER_1 = Counter.builder().id(ID_1).value(0).build();
    private static final List<Counter> COUNTER_LIST = Arrays.asList(COUNTER_1);

    @Mock
    GetRunningCountersUpdatesInteractor getRunningCountersUpdatesInteractor;
    @Mock
    GetCounterInteractor getCounterInteractor;
    @Mock
    UpdateCounterInteractor updateCounterInteractor;
    @Mock
    Timers timers;
    @Mock
    private RunningCounterPresenter.View view;
    @Captor
    private ArgumentCaptor<UiRunningCounterState> argumentCaptor;

    private UiCounterItemMapper uiCounterItemMapper = new UiCounterItemMapper();
    private PublishSubject<List<Counter>> subject;

    private RunningCounterPresenter sut;

    @Before
    public void setup() {
        subject = PublishSubject.create();
        given(getRunningCountersUpdatesInteractor.getRunningCountersUpdates()).willReturn(subject);
        given(updateCounterInteractor.updateCounter(COUNTER_1)).willReturn(Completable.complete());
        sut = new RunningCounterPresenter(getRunningCountersUpdatesInteractor,
                getCounterInteractor,
                updateCounterInteractor,
                uiCounterItemMapper,
                timers);
    }

    @After
    public void tearDown() {
        sut.destroy();
    }

    @Test
    public void firstAttachViewEmitsInitialState() {

        sut.attachView(view);

        verify(view).render(UiRunningCounterState.initialState());
    }

    @Test
    public void emitNewStateWhenCounterListUpdated() {
        sut.attachView(view);

        subject.onNext(COUNTER_LIST);

        verify(view, times(2)).render(argumentCaptor.capture());
        UiRunningCounterState uiRunningCounterState = argumentCaptor.getValue();
        assertEquals(1, uiRunningCounterState.getRunningItemList().size());
        assertEquals(ID_1, uiRunningCounterState.getRunningItemList().get(0).getId());
    }

    @Test
    public void startCounterStartsTimerIfNotRunning() {
        given(timers.hasTimer(ID_1)).willReturn(false);
        given(getCounterInteractor.getCounter(ID_1)).willReturn(Single.just(COUNTER_1));
        sut.attachView(view);

        sut.startCounter(ID_1);

        verify(timers, times(1)).startNew(eq(ID_1), eq(0L), any());
    }

    @Test
    public void startCounterDoesntStartTimerIfAlreadyRunning() {
        given(timers.hasTimer(ID_1)).willReturn(true);
        sut.attachView(view);

        sut.startCounter(ID_1);

        verify(getCounterInteractor, never()).getCounter(ID_1);
        verify(timers, never()).startNew(eq(ID_1), eq(0L), any());
    }

    @Test
    public void stopCounterStopsTimerIfRunning() {
        given(timers.hasTimer(ID_1)).willReturn(true);
        given(getCounterInteractor.getCounter(ID_1)).willReturn(Single.just(COUNTER_1));
        sut.attachView(view);

        sut.stopCounter(ID_1);

        verify(timers, times(1)).remove(ID_1);
    }

}

File: app/src/test/java/com/lucasurbas/counter/service/RunningCounterPresenterTest.kt
package com.lucasurbas.counter.service

import com.lucasurbas.counter.data.model.Counter
import com.lucasurbas.counter.service.usecase.GetCounterInteractor
import com.lucasurbas.counter.service.usecase.GetRunningCountersUpdatesInteractor
import com.lucasurbas.counter.service.usecase.UpdateCounterInteractor
import com.lucasurbas.counter.ui.explore.mapper.UiCounterItemMapper
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.capture
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.subjects.PublishSubject
import junit.framework.Assert.assertEquals
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.eq
import org.mockito.BDDMockito.given
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class RunningCounterPresenterTest {

    @Mock
    lateinit var getRunningCountersUpdatesInteractor: GetRunningCountersUpdatesInteractor
    @Mock
    lateinit var getCounterInteractor: GetCounterInteractor
    @Mock
    lateinit var updateCounterInteractor: UpdateCounterInteractor
    @Mock
    lateinit var timers: Timers
    @Mock
    lateinit var view: RunningCounterPresenter.View
    @Captor
    lateinit var argumentCaptor: ArgumentCaptor<UiRunningCounterState>

    val uiCounterItemMapper = UiCounterItemMapper()
    lateinit var subject: PublishSubject<List<Counter>>

    lateinit var sut: RunningCounterPresenter

    @Before
    fun setup() {
        subject = PublishSubject.create()
        given(getRunningCountersUpdatesInteractor.getRunningCountersUpdates()).willReturn(subject)
        given(updateCounterInteractor.updateCounter(COUNTER_1)).willReturn(Completable.complete())
        sut = RunningCounterPresenter(getRunningCountersUpdatesInteractor,
                getCounterInteractor,
                updateCounterInteractor,
                uiCounterItemMapper,
                timers)
    }

    @After
    fun tearDown() {
        sut.destroy()
    }

    @Test
    fun firstAttachViewEmitsInitialState() {

        sut.attachView(view)

        verify(view).render(UiRunningCounterState())
    }

    @Test
    fun emitNewStateWhenCounterListUpdated() {
        sut.attachView(view)

        subject.onNext(COUNTER_LIST)

        verify(view, times(2)).render(capture(argumentCaptor))
        val (runningItemList) = argumentCaptor.value
        assertEquals(1, runningItemList!!.size)
        assertEquals(ID_1, runningItemList[0].id)
    }

    @Test
    fun startCounterStartsTimerIfNotRunning() {
        given(timers.hasTimer(ID_1)).willReturn(false)
        given(getCounterInteractor.getCounter(ID_1)).willReturn(Single.just(COUNTER_1))
        sut.attachView(view)

        sut.startCounter(ID_1)

        verify<Timers>(timers, times(1)).startNew(eq(ID_1), eq(0L), any<Timers.TimerListener>())
    }

    @Test
    fun startCounterDoesntStartTimerIfAlreadyRunning() {
        given(timers.hasTimer(ID_1)).willReturn(true)
        sut.attachView(view)

        sut.startCounter(ID_1)

        verify<GetCounterInteractor>(getCounterInteractor, never()).getCounter(ID_1)
        verify<Timers>(timers, never()).startNew(eq(ID_1), eq(0L), any<Timers.TimerListener>())
    }

    @Test
    fun stopCounterStopsTimerIfRunning() {
        given(timers.hasTimer(ID_1)).willReturn(true)
        given(getCounterInteractor.getCounter(ID_1)).willReturn(Single.just(COUNTER_1))
        sut.attachView(view)

        sut.stopCounter(ID_1)

        verify<Timers>(timers, times(1)).remove(ID_1)
    }

    companion object {

        private val ID_1 = 1
        private val COUNTER_1 = Counter(ID_1, 0, false)
        private val COUNTER_LIST = Arrays.asList(COUNTER_1)
    }

}